### DISCUSSION TITLE
blog/the-query-options-api

### DISCUSSION THREAD

**Discussion by giscus** [TkDodo/blog-comments#102] [Announcements]
# blog/the-query-options-api

v5 brought a new, powerful API, especially if you're using React Query with TypeScript...

https://tkdodo.eu/blog/the-query-options-api#inject-comments

**Comment by lgenzelis** [comment]
That `dataTagSymbol` TS magic is pure gold! It allowed me to remove lots of unsafe generics from my codebase.

**Comment by danfry1** [comment]
Wow this is epic! Can't wait to start using `queryOptions` 🚀

**Comment by ginzatron** [comment]
Could you expand a little more on what the query-factory achieves and how it differs from "define keys in one central place while having the functions far a way from them in our custom hooks". What was the default patter before react query v5 and how has that changed with the options api? I'm having trouble fitting all the pieces together. Thanks.

**Reply by TkDodo** [reply]
In the post, I've linked to the article that explains how I did query key factories before:

https://tkdodo.eu/blog/effective-react-query-keys#use-query-key-factories

so, you'd just have a factory that creates keys, and then in your custom hook, you'd do:

```
const useTodo = (id) => useQuery({
  queryKey: todoKeys.detail(id),
  queryFn: () => fetchTodo(id),
})
```

**Comment by lucsky** [comment]
I've been using the query definition pattern for quite a while now and `queryOptions` will definitely help catch stupid mistakes, but I also have been using mutation definitions, in the same way, but there's no `mutationOptions` in sight. Why is that?

**Reply by TkDodo** [reply]
I haven't seen it as necessary to add mutationOptions. I mean if lots of people want it, we can add it, but all the advantages of the `queryOptions` api are basically gone:

- mutations are almost exclusively triggered by `useMutation`, so the only abstraction you need is a custom hook. There's not sharing the options between different functions.
- There's no need for the tagging functionality either

If you want type-safety, I think `satisfies MutationOptions` works well, too.

**Reply by abmagil** [reply]
I had a long running thread about how I wanted this in v4 so this is a wonderful discovery! 

I don't disagree about "necessary" but I ergonomics are a little weird that my factory is tied to the cache and the query, so I can just use the OptionsFactory attribute for the query but I can't use any of it for `useMutation` and then I can use _some_ of it for modifying my cache data. Example:

```typescript
const OptionsFactory = {
  ...
  entity: (id: Entity["id"]) => ({queryKey: ['entities', id], queryFn: myGetFn})
}

const { data: entity } = useQuery(OptionsFactory.entity(someId))

const create = useMutation({mutationFn: myPostFn})
const update = useMutation({mutationFn: myPutFn})

create.mutate({onSuccess: (returnedEntity) => queryClient.setCacheData(OptionsFactory.entity(returnedEntity.id) })
```

The ergonomics are even more pronounced if I were to use `onSuccess` inside the hook's options, instead of the `.mutate`'s options.

Regardless, since I'd want to PUT and POST to different URLS (`PUT /entities/{id}` and `POST /entities/` for updating and creating, respectively), I'm not sure how to best organize my OptionsFactory to support the various use cases.

**Comment by cobyme** [comment]
You have ads for Query.gg all over the place, but no way to actually purchase it yet. 😢 Eagerly awaiting that email & Tweet saying it's ready and available to purchase.

**Reply by TkDodo** [reply]
Hey 👋 
I'm working as hard as I can on query.gg, but I only do this part-time. I have a job and a family, too :)

For now, please subscribe to the mailing list and I'll make sure to inform you as soon as we have the first lessons out 🚀

**Comment by matzeso** [comment]
Great to see typescript improvements coming with v5.
Having read your examples before we also built our newest product using react router + react query with abstracted query definitions, and we also handled the type safety situation by building our own query util function:

```
import { QueryKey, FetchQueryOptions } from '@tanstack/react-query';

export const createQuery = <
  TQueryFnData = unknown,
  TError = unknown,
  TData = TQueryFnData,
  TQueryKey extends QueryKey = QueryKey,
>(
  options: FetchQueryOptions<TQueryFnData, TError, TData, TQueryKey>,
) => options;
```

I guess we can replace that after migrating to v5 with the new `queryOptions` helper :)

**Comment by rlesniak** [comment]


**Reply by TkDodo** [reply]
Because it's a different thing? You want infiniteQueryOptions

**Comment by rudfoss** [comment]
This is a great addition to an already great library! I'm wondering if you have a recommended pattern for when context provided clients are used (or maybe I'm using it wrong to begin with)?

I'm using generated clients (NSwag) that I provide to my application through contexts. This has served me well as an abstraction and allows me to relatively easily set up clients for different environments and use them in my application.

My custom query hooks look something like this:
```ts
export const useUserInformation = (userId: string) => {
  const { userInformationClient } = useUserInformationClient()
  return useQuery({
    queryKey: ["users", "byId", id],
    queryFn: () => userInformationClient.getUserInformation(userId),
    ... // other options
  })
}
```

From my components I can then simply use the hook without knowing about the underlying dependency:

```ts
const { data: userInformation } = useUserInformation(userId)
```

With `queryOptions` it seems I would have to expose the client in every component that uses the query or maybe capturing the client in a closure or something. How would you recommend I encapsulate the client instance? Or perhaps I should rethink how I structure my "API layer"?

Pure queryOptions:
```ts
const userInformationQueries = {
  byId: (client: UserInformationClient, userId: string) => queryOptions({
    queryKey: ["users", "byId", id],
    queryFn: () => client.getUserInformation(userId),
    ... // other options
  })
}

// Usage from a component (works, but must repeat a lot of stuff)
const { userInformationClient } = useUserInformationClient()
const { data: userInformation } = useQuery(userInformationQueries.byId(userInformationClient, userId))
```

In closure (seems like a better approach if combined with context):
```ts
const userInformationQueries = (client: UserInformationClient) => ({
  byId: (client: UserInformationClient, userId: string) => queryOptions({
    queryKey: ["users", "byId", id],
    queryFn: () => client.getUserInformation(userId),
    ... // other options
  })
})

// In a component (could be abstracted to a separate context and provided)
const { userInformationClient } = useUserInformationClient()
const { data: userInformation } = useQuery(userInformationQueries(userInformationClient).byId(userId))

// Or if provided from context (still have multiple lines, but at least it hides the client dependency):
const { byId } = useUserInformationQueries()
const {data: userInformation } = useQuery(byId(userId))
```

**Reply by rudfoss** [reply]
(seems i forgot to remove the client argument from `byId` (line 2) in the last code-block. I can't seem to be able to edit my post so please ignore that argument).

**Reply by TkDodo** [reply]
I don't think its generally a good idea to have the client live in react context - what's the advantage here? If you're using our eslint plugin, the exhaustive-deps rule will also flag that `userInformationClient` is used in the queryFn, but it isn't part of the key. You can technically run into stale closure issues if your client ever gets re-created, because the queryFn might not read the latest value then.

So yeah, you avoid all of that if the client doesn't live in react context at all, but is just a stable value that you import.

**Reply by rudfoss** [reply]
The primary advantage I see is that I'm using the same mechanism to "inject" dependencies for everything. This makes it easier to follow the lifecycle flow of code in the application (IMHO) and configurations can be passed through props just like any other component.

The idea is not that clients vary greatly while the application lives. They are usually instantiated with some configuration on startup and only updated if the user logs in and a token needs to be added to the client headers or some such process. It's rather that I can use one mechanism to pass them around my application along with any other dependencies just like the Tanstack Query Provider provides a centrally configured client. Using contexts means there is no "custom mechanism" to learn for other developers when working with clients and it follows the same lifecycle as any other react component.

I can't seem to agree with myself whether it is cleaner than introducing a "management" mechanism for clients outside of the react tree, hence why I was asking, but it does make for a type-safe and relatively clean API when using the queries and mutations. It will cause some extra re-renders that could be avoided, perhaps by using refs, but at least it follows pre-existing react render- and lifecycle flows.

```tsx
const createApiClientQueries = (apiClient: IApiControllerClient) => {
  const queries = {
    users: () => ["users"],
    byId: (id: string) => queryOptions({
      queryKey: [...queries.users(), "byId", id]
      queryFn: () => apiClient.getUser(id),
      staleTime: 60_000
    })
  } as const
  return queries
}

// assume the context is defined to contain what is provided in `value`
const ProvideApiClient = ({baseUrl, children}: {baseUrl: string, children: ReactNode) => {
  const [token, setToken] = useState<string>("")
  const apiClient = useMemo(() => new ApiControllerClient(baseUrl, token), [baseUrl, token])
  const queries = useMemo(() => createApiClientQueries(apiClient), [apiClient])

  return (
    <ApiClientContext.Provider value={{apiClient, queries, token, setToken}}>{children}</ApiClientContext.Provider />
  )
}

const UsersPage = () => {
  const { queries, token } = useContext(ApiClientContext)
  const { data } = useQuery({...queries.getUsers(), enabled: !!token })
  const { mutate: login } = useLogin() // Custom mutation hook following suggested pattern

  if (data) {
    return <UsersTable users={data} />
  }

  return <LoginForm onLogin={login} />
}

const App = () => (
  <ProvideApiClient baseUrl="http://localhost:1234">
    <UsersPage />
  </ProvideApiClient>
)
```

**Reply by TkDodo** [reply]
the problem is that APIClient depends on token, and if token changes, API client changes. Then, in your `queryFn`, you call `apiClient.getUser`, which will then e.g. add the current token as a header parameter.

But since `token` isn't part of your `queryKey`, when you invalidate a query, it might read an older queryFn, where the apiClient still has an older token. That's a stale-closure issue. I have a blogpost on this topic: https://tkdodo.eu/blog/hooks-dependencies-and-stale-closures

Our lint rule warns against that, but the solution to add `apiClient` to the key is wrong unless the client is serializable. Also, you don't really want separate cache entries just because the token changed.

So the solution is to imperatively _read_ the token rather than declaratively depend on it. Like, put it in a ref or a zustand store and in the queryFn call `getState().token` or whatever

**Reply by rudfoss** [reply]
ah, sorry. I didn't fully get the queryFn-caching mechanism. That would indeed make this impractical and hide a potentially big problem in a closure... A zustand store would still introduce a new mechanism though as I cannot use the store-hook to create queries so I would have to call it directly.

The issue as I see it still remains though. We have an "injection mechanism" with prop-management in Contexts that is built into React. I want to "inject" dependencies (clients) so they can be used by my components, and manage their lifecycles so that changes that require a new client propagate properly. The client is a dependency the queryFn needs to deal with, but it cannot use the built in dependency injection mechanism or indeed any state mechanism to do so requiring some workaround or external management mechanism (or refs) to work correctly.

I understand your reasoning though and it makes perfect sense, though I'm conflicted in trying to keep it simple by avoiding a new mechanism or by introducing unnecessary wrappers. It seems "correct" to me that parameters provided as props (e.g. baseUrl, tokens etc) should be allowed to follow the lifecycle of react throughout my application. There is probably something wrong with my underlying assumptions here. In any case thanks for replying here. I really appreciate the help! I'll take a step back and grab an extra 🍵 and see if I can rework it from the top to fit with your recommendations.

**Reply by TkDodo** [reply]
> The client is a dependency the queryFn needs to deal with, but it cannot use the built in dependency injection mechanism or indeed any state mechanism to do so requiring some workaround

what you can do is:
- add the client to the queryKey
- implement a custom `queryKeyHashFn` (globally)
  - serialize the client to some static value, e.g. an id (by using `client.getId()` or so)
  - serialize everything else with our default `hashKey` fn (you can import it)
- make sure that id increases / changes every time the client is re-created, like when a new token is issued or baseUrl changes.

that will satisfy the linter and will not have any stale closure problems, while still allowing you to use react-context for dependency injection

**Comment by hrachhovsepyanqless** [comment]
Thanks for the article, great read! How would you recommend to handle the possibility to add (extend) queryOptions for each query? Let's say I want to disable a certain query in some scenarios, is there any pattern here that handles it in the query factory?

**Reply by TkDodo** [reply]
you could just pass it as an argument to the function of the factory, or you'd just spread it onto the useQuery call where you need it:

```
const useMyQuery = () => useQuery({
 ...myQueryOptions(),
 enabled: someCondition
})
```

**Reply by gerardosabetta** [reply]
So what if you wanted to make this super flexible but also type safe? is there a way to do 

```
const useMyQuery = (config) => useQuery({
 ...myQueryOptions(),
 ...config
})
```

while making that `config` parameter type safe? I'm thinking maybe I don't want to create `useMyQueryCount` which would be basically the same thing with a `select` implementation on it. How would you type this `config` object?

**Reply by TkDodo** [reply]
I always recommend against doing that. Getting the types 100% right for this, you'd need to implement basically what `useQuery` is doing, which means [4 type parameters and 3 overloads](https://github.com/TanStack/query/blob/03048e3776041ff5f5776b1b739c18183aa7e33a/packages/react-query/src/useQuery.ts#L15-L43). It's not worth it. This abstraction doesn't do anything when `queryOptions` are involved. Suppose I'd want to call:

```
const { data } = useMyQuery({ refetchInterval: 1000 })
```

I could just as well do:

```
const { data } = useQuery({
  ...myQueryOptions(),
  refetchInterval: 1000
})
```

you need absolutely zero type shenanigans for this. It'll be 100% type safe, even when `select` is involved. And you can pass that to `useQuery` or `useSuspenseQuery` or `useQueries` or `queryClient.prefetchQuery` and so on ... `queryOptions()` is by far the best and likely only abstraction you need to make things type-safe AND super flexible.

**Comment by Thanaen** [comment]
Thanks for this great article!
I'd love to implement it, but I'm wondering something:
In my application, all requests systematically take a parameter from a store (in this case, zustand). It's useful for determining the URL of the API call, and it's passed in all react-query keys.

What would be the best way to set up a query factory in this situation? I'm thinking that systematically retrieving the value as a parameter from the factory would be very cumbersome, and it would force each component to retrieve the value from the store before calling the query.

I'm thinking I could do a wrap with, say, useUserQuery and useSuspenseUserQuery, which take care of retrieving the value from the store and injecting it, but Typescript won't be happy about that, as I'm using the QueryFunctionContext to determine the parameters of my API helpers...
Do I need to make a wrapper beyond "queryOptions", which automatically adds the "userId" (in my example) to the list of queryKeys?

Sorry if I'm asking a lot, but I'm sure you've come across this situation before 😄

Once again, thank you very much for your work!

**Reply by Thanaen** [reply]
Follow up on what I've done so far:

I created a "userQueries" function like this:
```
export const userQueries = (userId: string) => ({
  getUserDetails: () =>
    queryOptions({
      queryKey: ["userDetails", { userId }],
      queryFn: getUserDetails,
    }),
});
```

Then I created a context like this: 
```
const userQueriesContext = React.createContext<ReturnType<
  typeof userQueries
| null>(null);

export const UserQueriesProvider = ({
  children,
}: ServerQueriesProviderProps) => {
  const userId = useStore(currentUserIdSelector);

  const computedUserQueries = React.useMemo(
    () => userQueries(userId),
    [userId],
  );

  return (
    <userQueriesContext.Provider value={computedUserQueries}>
      {children}
    </userQueriesContext.Provider>
  );
};
```

I don't think it's too bad!

**Reply by Thanaen** [reply]
Then all I have to do is call this hook in my components, and voila!
```
const useUserQueries = () => {
  const context = React.useContext(userQueriesContext);
  if (!context) {
    throw new Error(
      "useUserQueries must be used within a UserQueriesProvider",
    );
  }
  return context;
};
```

This also allows me to invalidate / interact with queries outside of React components, since I can easily call my "userQueries" function by passing it the userId by hand!

**Reply by TkDodo** [reply]
I'd make a custom hook:

```
export const useUserDetailsOptions = () => {
  const userId = useStore(currentUserIdSelector)

  return queryOptions({
    queryKey: ["userDetails", { userId }],
    queryFn: getUserDetails,
  })
}
```

**Comment by Arcades2** [comment]
So cool as always ! For the query factory would you still recommend the queryKey as an array of one object like in your article "Leveraging the Query Function Context" ? I love the fact that it forces me to put in the queryKey what I need in my queryFn. That way, there's less chance of misidentifying my query.

**Reply by TkDodo** [reply]
yeah I still do that, but it doesn't matter as much if key + fn are colocated

**Comment by wuarmin** [comment]
Hey,
just one question about prefetching in the context of tanstack Router: Which react query function do you recommend for prefetching in the route (loader-fn)? I keep seeing `ensureQueryData` in the router examples. Here you mention `prefetchQuery`. What is the right one? And why?

Thanks for your great Work!

**Reply by TkDodo** [reply]
I prefer `ensureQueryData` because of the reasons outlined here:
- https://tkdodo.eu/blog/react-query-meets-react-router#getquerydata--fetchquery

**Comment by dreyescabrera** [comment]
Thanks a lot Dom. Couldn't we do something similar with mutations?

**Reply by TkDodo** [reply]
There isn't really a need for something similar for mutations because:

- they have no key, so the `dataTag` association between key / fn isn't necessary.
- you can thus just make a function that returns something that `satisfies MutationOptions` and you're good to go.

**Reply by shkumbinhasani** [reply]
I disagree, on many cases the onSuccess inference is very helpful and can contain a lot of logic, which doesnt happen on satisfies MutationOptions, it would be a nice utility to have, in the side of queryOptions

**Reply by TkDodo** [reply]
yeah, I might be coming around on this one :)

**Comment by markbenliyan** [comment]
If we still wanted to use the QueryFunctionContext pattern outlined here, https://tkdodo.eu/blog/leveraging-the-query-function-context how could we get type the queryKey for the examples like todoQueries.list or todoQueries.detail?

**Reply by TkDodo** [reply]
query key factories and option factories are not really different. Instead of producing a key, they just produce a full object, and you can combine this with the QueryFunctionContext pattern like so:

```
const todoQueries = {
  all: () => [{ entity: 'todos' }],
  lists: () => [{ ...todoKeys.all[0], entity: 'list' }] as const
  list: (filters: string) =>
    queryOptions({
      queryKey: [{ ...todoKeys.lists()[0], filters }] as const,
      // ✅ extract named properties from the queryKey
      queryFn: ({ queryKey: [{ filters }]}) => fetchTodos(filters),
    }),
}
```

However, I think extracting properties from the QueryFunctionContext like that isn't really necessary in the setting of a query options factory. `filters` are in the scope already, and our lint rule from eslint-plugin-query will make sure that we don't use anything that isn't part of the key. The problem that this pattern was trying to solve was largely eliminated with the lint rule + the query options API. As you can see, you can combine them, but I don't think it's worth it.

**Comment by AlexKMarshall** [comment]
I'm really enjoying using queryOptions for sharing the query among various functions running on the client. But, I'm struggling to see how this will fit into server-side prefetching. We're using Next pages router, with getServersideProps, but I think the issue would be the same with Remix/React Router loaders too. We're using the Next server functions as a BFF. When the client fetches it does so against a proxy api route that forwards requests on to the main backend API.

The auth between the client and the BFF is an http only cookie storing a jwt. In the BFF proxy that's unwrapped and added as an auth header before sending to the backend API, so that all works fine. The fetcher functions in queryOptions don't need to know anything about auth.

```ts
const orderQueryOptions = (orderId) => queryOptions({queryKey: ['orders', orderId], queryFn: () => fetchOrder(orderId)})
const orderQuery = useQuery(orderQueryOptions(orderId))
```

But for the getServersideProps calls we've got to read the request cookies to get the access token and add that to the fetcher. I can't see how to do that without adding the access token in a closure to the query options. But then it can't be part of the query key and run the risk of a stale closure

```ts
const orderQueryOptions = (orderId: string, accessToken?: string) => queryOptions({queryKey: ['orders', orderId], queryFn: () => fetchOrder(orderId, accessToken)})
// in getServersideProps
const accessToken = getToken(request.cookies)
queryClient.prefetchQuery(orderQueryOptions(orderId, accessToken))
```

In another comment thread you've suggested that the fetcher functions should imperatively get things like tokens from a store or ref, so they don't need to be in the query key. This would work if we were only fetching client-side, but I don't think that can work on the server which needs to remain stateless to avoid leaking data between users.

Is this something that you have a pattern for?

**Reply by TkDodo** [reply]
I think you would want different QueryFunctions on the client and on the server

**Reply by AlexKMarshall** [reply]
Yeah, that seems like the only way to go, though I guess you then lose some of the benefits of the queryOptions and the linter

```ts
const defaultOrderQueryOptions =(filters) =>  queryOptions({
  queryKey: ['orders', filters], // This will warn if I miss a dependency in the key
  queryFn: () => fetchOrders(filters) 
})
// In a component
const filters = useFilters()
const ordersQuery = useQuery(defaultOrderQueryOptions(filters))
// On the server
const {filters, sort} = request.queryParams;
const accessToken = getAccessToken(request.cookie)

queryClient.prefetchQuery({
  ...defaultOrderQueryOptions(filters),
  queryFn: () => fetchOrdersWithAuth(accessToken, filters, sort) // I won't get a warning from the linter that I'm passing a sort param that's not in the key
})
```

And I could easily end up fetching a completely different resource by mistake, one of the problems I'm trying to solve with `queryOptions` in the first place, (colocating the fetch function with the key, especially if they have more complex logic in them)

```ts
queryClient.prefetchQuery({
  ...defaultOrderQueryOptions(filters),
  queryFn: () => fetchCartWithAuth(accessToken) 
})
```

**Comment by JoakimGit** [comment]
Great article Dodo. I've been using the pattern in my current project, but have a question about file location. This is the bookingQueries object I export and use in my components:

```
// ../features/booking/api/booking-queries.ts
export const bookingQueries = {
  all: () => ['booking'],
  calendar: () => [...bookingQueries.all(), 'calendar'],
  bookablePeriods: (propertyId: string) => queryOptions({
    queryKey: [...bookingQueries.calendar(), propertyId, 'bookablePeriod'],
    queryFn: () => fetchBookablePeriods(propertyId),
  }),
  bookings: (propertyId: string) => queryOptions({
    queryKey: [...bookingQueries.calendar(), propertyId, 'booking'],
    queryFn: () => fetchBookings(propertyId),
  }),
  firstAndLastPeriodByDestination: (destinationId: string | undefined) => queryOptions({
    queryKey: [...bookingQueries.all(), 'bookablePeriod', 'firstAndLast', destinationId],
    queryFn: destinationId ? () => fetchFirstAndLastBookablePeriod(destinationId) : skipToken,
    select: ({ earliestDateFrom, latestDateTo }) => ({
      calendarStart: dayjs(earliestDateFrom ?? undefined).startOf("isoWeek"),
      calendarEnd: dayjs(latestDateTo ?? undefined).endOf("isoWeek")
    }) as CalendarBoundary,
  }),
  period: (periodId: string | undefined) => queryOptions({
    queryKey: [...bookingQueries.all(), 'period', periodId],
    queryFn: periodId ? () => fetchPeriod(periodId) : skipToken,
  }),
  companyBooking: (bookingId: string | undefined) => queryOptions({
    queryKey: [...bookingQueries.all(), 'company', bookingId],
    queryFn: bookingId ? () => fetchCompanyBooking(bookingId) : skipToken,
  }),
  userBooking: (bookingId: string) => queryOptions({
    queryKey: [...bookingQueries.all(), 'user', bookingId],
    queryFn: () => fetchUserBooking(bookingId),
  }),
  companyBookingTypes: () => queryOptions({
    queryKey: [...bookingQueries.all(), 'companybookingtypes'],
    queryFn: fetchCompanyBookingTypes,
  }),
}
```

```
const BookingCalendarRow = ({ property }: BookingCalendarRowProps) => {
  const bookingsQuery = useQuery(bookingQueries.bookings(property.id));
  // rest of component
}
```

My question is regarding where you place the actual fetch functions and potentially any custom hooks. E.g. 

```
export const useCalendarBoundary = () => {
  const destination = useSelectedDestination();
  return useQuery(bookingQueries.firstAndLastPeriodByDestination(destination?.id));
}

export const useCalendarRange = () => {
  const boundaryQuery = useCalendarBoundary();
  const { calendarStart, calendarEnd } = boundaryQuery.data || {};

  if (!calendarStart || !calendarEnd) return 0;

  return calendarEnd.diff(calendarStart, "days") + 1; // +1 to include both first and last day
}
```

1. Do you keep it all in the same file? 

2. One file for the bookingQueries object, another for the fetch functions, a third for hooks?

3. A file for each query, that contains the fetch function, the queryOptions object and any potential custom hooks, which is then all imported by bookingQueries.ts?
4. ??

I know it's more a question of opinion, but I'd like to hear some thoughts.


Another thing not related to this. The project uses TanStack Router. I'm doing queryClient.ensureQueryData in the loader. Then in the component, I'm not sure when to use useQuery or useSuspenseQuery.

**Comment by ak274** [comment]
From the query factory lets say if i only want to invalidate a specific detail By Id how can i do that?

```queryClient.invalidateQueries({queryKey: keys.details()});``` will invalidate all the details by id  fetched previously.

So as per your example I only see the way is to explicitly give id like this 
```queryClient.invalidateQueries({queryKey: [...keys.details(), id ]});``` which i feel is not good way as now I'm in a way creating a new query key.

**Reply by TkDodo** [reply]
if you use the `todoQueries` factory as described in the article at the end, you would do:

```
queryClient.invalidateQueries(todoQueries.details(5))
```

this would produce the key `['todos', 'detail', 5]` and would therefore only match that specific key.

**Reply by ak274** [reply]
But in example `todoQueries.details` doesn't take any argument .
 
`todoQueries.detail` has a id argument but its is with queryOption  object so not sure if I can invalidate with this?

**Reply by TkDodo** [reply]
right, sorry, just a typo on my end:

- `queryClient.invalidateQueries(todoQueries.detail(5))` will invalidate only todo 5
- `queryClient.invalidateQueries(todoQueries.details())` will invalidate all todo details

> but its is with queryOption object so not sure if I can invalidate with this?

yes, you can.

**Reply by ak274** [reply]
I was not able to invalidate like this:

`queryClient.invalidateQueries(todoQueries.detail(5))`

I'm getting the folloing error:

```Type 'UseQueryOptions<AxiosResponse<string[]>, Error, AxiosResponse<string[]>, string[]> & { initialData?: InitialDataFunction<AxiosResponse<string[]>> | undefined; } & { ...; }' is not assignable to type 'QueryKey | undefined'.```

instead of it I had to append the queryKey property:

`queryClient.invalidateQueries(todoQueries.detail(5)).queryKey`

**Reply by ak274** [reply]
Sorry the above mentioned way to invalidate Queries works as expected but the issue I'm having is with  

`queryClient.setQueryData`

Now If i want to explicitly  setQueryData to `todoQueries.details()` what should be the approach? The following code won't work as first argument expect a query key but I'm providing a queryObject.

```queryClient.setQueryData<Todo>(
        todoQueries.detail(id),
        (old) => ({
          ...old,
          ...data,
        }),
      );```

**Reply by TkDodo** [reply]
You pass `todoQueries.detail(id).queryKey` as the first argument

**Comment by AnhV90** [comment]
Love using queryOptions :heart:

I have one question tho. I also love using [Query Key Factory](https://github.com/lukemorales/query-key-factory) but it seems queryOptions is overlapping some features with QKF. So how should I combine both to have nicer query key object (without having spread the lists into list's queryKey and with `_def` lists can be removed) but also want to keep the queryOptions for reuse purpose?

**Reply by TkDodo** [reply]
we’d have to ask @lukemorales about what his plans are to make use of `queryOptions` with his lib.

**Comment by Omer-Mazig** [comment]
This is mind blowing! 🤯 

This all blog series is pure gold, but this blog is the the the cream of the crop.

Thank you for your hard work on this amazing library 💕

**Comment by ryoid** [comment]
When using query options factory, is the recommended way to access `queryClient` via argument?

For example to pull initial data
```tsx
const TodoQueries = {
  list: () => {
    return queryOptions({
      queryKey: ['todos', 'list'],
      queryFn: () => fetchTodos(),
    });
  },
  detail: (id: number, queryClient: QueryClient) => {
    return queryOptions({
      queryKey: ['todos', 'detail', id],
      queryFn: () => fetchTodo(id),
      initialData: () => {
        return queryClient.getQueryData(TodoQueries.list().queryKey)?.find((todo) => todo.id === id);
      },
      initialDataUpdatedAt: () => queryClient.getQueryState(TodoQueries.list().queryKey)?.dataUpdatedAt,
    });
  },
};
```

**Reply by TkDodo** [reply]
you can also make it a hook itself and call `useQueryClient` inside.

**Comment by vaziliybober** [comment]
Hey! In your article you have a statement "The queryKey defines the dependencies to our queryFn - everything we use inside it must go into the key". But what if it's something like an http client? For example, I might have an http client that I explicitly pass to my queryOptions builder to be used inside the queryFn. But I don't need it for the queryKey.

Another example would be the queryClient, which can sometimes be used to put some initialData in the queryOptions for example, but the queryKey doesn't care about it.

And now I can't generate my queryKeys without dependencies like an http client or queryClient.

Do you think it's better to stick with a separate factory for keys only then?

**Reply by TkDodo** [reply]
if it’s a static that you import, it won’t be reported. If it’s something that you get from `useHttpClient()`, there is a chance that it’s dynamic, like, idk, changing base-url or so, and then it would need to somehow go into key (you’d need a way to serialize it).

> Another example would be the queryClient

queryClient is passed into the queryFn as `client`:

```
queryFn: ({ client}) => ...
```

**Reply by vaziliybober** [reply]
> there is a chance that it’s dynamic

In my case it is only dynamic when it comes to its side effects. So it is deterministic, but side effects can be different (like logging differently, for example).

**Comment by Vesper0704** [comment]
Thanks for your sharing. The `query-factory` part is inspiring. 
However, I'm using Next.js and client/server request differs due to some `server-only` import. 
In these cases, is it possible to organize queryFn and queryKeys in the single file as you demoed above?

**Reply by TkDodo** [reply]
I have no idea. What’s server-only in your queryOptions ?

**Reply by Vesper0704** [reply]
queryFn may include fetch that needs `next/headers` for auth, and importing them in client components throw Error.

**Reply by TkDodo** [reply]
you can’t use something that works on the server only on the client 🤷‍♂️

**Comment by yeunoia** [comment]
what about when using `mutaionOption`?
would you define all together in factories? with queryOptions and mutationOptions all together?
 
```ts
const todoQueries = {
  detail: (id: number) =>
    queryOptions({
      queryKey: [...todoQueries.details(), id],
      queryFn: () => fetchTodo(id),
      staleTime: 5000,
    }),
  mutateDetails : (id: number) => 
   mutationOptions({
      mutateFn: () => fetchTodo(id),
     })
}
```

**Reply by TkDodo** [reply]
that’s really up to you

**Comment by LovesWorking** [comment]
Isn't the query key factory still better as you can call one function in multiple spots to invalidate queries instead of having to write strings?

**Reply by TkDodo** [reply]
They are not mutually exclusive. You can make option factories like shown here: 

- https://tkdodo.eu/blog/the-query-options-api#query-factory

**Comment by ptphongkmf** [comment]
Hey, thanks for the great post. I have a quick question about the query options factory pattern. If we group all our keys and fetch functions into one big exported object, wont that mess up code splitting? Like if i import just the list query on my homepage, it seems like the bundler will pull all the other queries and their utils into the homepage chunk, since they are attached to the same object export. Maybe im overthinking the perf problem. In your experience, is the impact negligible enough for the dx?

**Reply by TkDodo** [reply]
I wouldn’t do one big object - I’d do one object per feature.

Just pulling in the query objects (key + fn + options) for one feature is usually not the problem. They usually all hit the same `api` client for making a request, which is in the critical path anyways, so yes, I think you’re overthinking it a bit.
